wfe2/wfe.go

```diff
@@ -1486,21 +1539,145 @@ func (wfe *WebFrontEndImpl) Order(ctx context.Context, logEvent *web.RequestEven
+			probs.Unauthorized("Must agree to subscriber agreement before any further actions"), nil)
+		return
+	}
+
+	// If the order's status is not pending we can not finalize it and must
+	// return an error
+	if *order.Status != string(core.StatusPending) {
+		wfe.sendError(response, logEvent,
+			probs.Malformed("Order's status (%q) was not pending", *order.Status), nil)
+		return
 	}
 
+	// If the order is expired we can not finalize it and must return an error
+	orderExpiry := time.Unix(*order.Expires, 0)
+	if orderExpiry.Before(wfe.clk.Now()) {
+		wfe.sendError(response, logEvent, probs.NotFound(fmt.Sprintf("Order %d is expired", *order.Id)), nil)
+		return
+	}
+
+	// The authenticated finalize message body should be an encoded CSR
+	var rawCSR core.RawCertificateRequest
+	err := json.Unmarshal(body, &rawCSR)
+	if err != nil {
+		wfe.sendError(response, logEvent,
+			probs.Malformed("Error unmarshaling finalize order request"), err)
+		return
+	}
+
+	// Check for a malformed CSR early to avoid unnecessary RPCs
+	csr, err := x509.ParseCertificateRequest(rawCSR.CSR)
+	if err != nil {
+		wfe.sendError(response, logEvent, probs.Malformed("Error parsing certificate request: %s", err), err)
+		return
+	}
+
+	certificateRequest := core.CertificateRequest{Bytes: rawCSR.CSR}
+	certificateRequest.CSR = csr
+	wfe.logCsr(request, certificateRequest, *acct)
+
+	// Check that the key in the CSR is good. This will also be checked in the CA
+	// component, but we want to discard CSRs with bad keys as early as possible
+	// because (a) it's an easy check and we can save unnecessary requests and
+	// bytes on the wire, and (b) the CA logs all rejections as audit events, but
+	// a bad key from the client is just a malformed request and doesn't need to
+	// be audited.
+	if err := wfe.keyPolicy.GoodKey(certificateRequest.CSR.PublicKey); err != nil {
+		wfe.sendError(response, logEvent, probs.Malformed("Invalid key in certificate request :: %s", err), err)
+		return
+	}
+	logEvent.Extra["CSRDNSNames"] = certificateRequest.CSR.DNSNames
+	logEvent.Extra["CSREmailAddresses"] = certificateRequest.CSR.EmailAddresses
+	logEvent.Extra["CSRIPAddresses"] = certificateRequest.CSR.IPAddresses
+
+	// Inc CSR signature algorithm counter
+	wfe.csrSignatureAlgs.With(prometheus.Labels{"type": certificateRequest.CSR.SignatureAlgorithm.String()}).Inc()
+
+	err = wfe.RA.FinalizeOrder(ctx, &rapb.FinalizeOrderRequest{
+		Csr:   rawCSR.CSR,
+		Order: order,
+	})
+	if err != nil {
+		wfe.sendError(response, logEvent, web.ProblemDetailsForError(err, "Error finalizing order"), err)
+		return
+	}
+	// Convert the corepb.Order.Names into identifiers for display to the user.
+	idents := make([]core.AcmeIdentifier, len(order.Names))
+	for i, name := range order.Names {
+		idents[i] = core.AcmeIdentifier{Type: core.IdentifierDNS, Value: name}
+	}
+
+	respObj := wfe.orderToOrderJSON(request, order)
 	err = wfe.writeJsonResponse(response, logEvent, http.StatusOK, respObj)
```

Comment:
Will this not be missing the certificate serial and still have a pending status since `ra.FinalizeOrder` isn't mutating `order` or am I missing something?

Reply:
That's correct. The spec doesn't say that the finalize message should return the order with the certificate serial populated. The client is meant to poll the order after POSTing the finalize endpoint so that the Order can be issued asynchronously from the finalize request.

I _think_ we should probably be mutating the order state to "Processing" but this will require an adjustment to the SA's Order RPCs since there isn't a way to update just the status field of an order with the current Order RPCs.

Reply:
Maybe since we are issuing synchronously anyway we should just skip the processing state and change the RA to return the certificate serial and the order in the valid state.

Reply:
We will want to switch to async issuance pretty soon as part of the SCT embedding work, so we should be planning ahead for that. However, I think it would probably make sense for that to be its own RPC: SetOrderProcessing or similar.

Reply:
Sounds good. I'll have the RA set the order to status processing with a dedicated RPC and the WFE can return the order without the CertificateSerial but with Status: Pending at the end of a successful finalize order request.